arch/risc-v/src/k230/k230_rptun.c

```diff
@@ -0,0 +1,501 @@
+      rsc->rsc_tbl_hdr.reserved[0]  = 0;
+      rsc->rsc_tbl_hdr.reserved[1]  = 0;
+      rsc->offset[0]                = offsetof(struct rptun_rsc_s,
+                                               rpmsg_vdev);
+
+      rsc->rpmsg_vdev.type          = RSC_VDEV;
+      rsc->rpmsg_vdev.id            = VIRTIO_ID_RPMSG;
+      rsc->rpmsg_vdev.dfeatures     = 1 << VIRTIO_RPMSG_F_NS
+                                    | 1 << VIRTIO_RPMSG_F_ACK
+                                    | 1 << VIRTIO_RPMSG_F_BUFSZ;
+      rsc->rpmsg_vdev.config_len    = sizeof(struct fw_rsc_config);
+      rsc->rpmsg_vdev.num_of_vrings = VRINGS;
+
+      rsc->rpmsg_vring0.align       = VRING_ALIGN;
+      rsc->rpmsg_vring0.num         = VRING_NR;
+      rsc->rpmsg_vring0.notifyid    = VRING0_NOTIFYID;
+      rsc->rpmsg_vring1.align       = VRING_ALIGN;
+      rsc->rpmsg_vring1.num         = VRING_NR;
+      rsc->rpmsg_vring1.notifyid    = VRING1_NOTIFYID;
+      rsc->config.r2h_buf_size      = VRING_SIZE;
+      rsc->config.h2r_buf_size      = VRING_SIZE;
+
+      priv->shmem->base             = (uintptr_t)priv->shmem;
+
+      rpinfo("shmem:%p, dev:%p\n", priv->shmem->base, dev);
+    }
+  else
+    {
+      /* TODO: use IPI later, polling now. */
+
+      rpinfo("wait for shmem %p...\n", priv->shmem);
+      while (priv->shmem->base == 0)
+        {
+          nxsig_usleep(100);
+        }
+
+      rpinfo("ready to go!\n");
+    }
+
+  return &priv->shmem->rsc;
+}
+
+/****************************************************************************
+ * Name: rp_is_autostart
+ ****************************************************************************/
+
+static bool rp_is_autostart(struct rptun_dev_s *dev)
+{
+  return true;
+}
+
+/****************************************************************************
+ * Name: rp_is_master
+ ****************************************************************************/
+
+static bool rp_is_master(struct rptun_dev_s *dev)
+{
+  struct k230_rptun_dev_s *priv = as_k230_rptun_dev(dev);
+  return priv->master;
+}
+
+/****************************************************************************
+ * Name: rp_start
+ ****************************************************************************/
+
+static int rp_start(struct rptun_dev_s *dev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rp_stop
+ ****************************************************************************/
+
+static int rp_stop(struct rptun_dev_s *dev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rp_notify
+ ****************************************************************************/
+
+static int rp_notify(struct rptun_dev_s *dev, uint32_t vqid)
+{
+#ifdef CONFIG_K230_RPTUN_IPI
+  k230_ipi_signal(RPTUN_IPI_NOTIFY_RX);
+#else
+  struct k230_rptun_dev_s *priv = as_k230_rptun_dev(dev);
+
+  if (priv->master)
+    {
+      priv->shmem->seq_rmt++;
```

Comment:
don't need seq_rmt/seq_mst.

Reply:
They are still needed for this non-IPI based solution., later once IPI is available, they will be turned off by MACROS.